Speed up the ContainExactly matcher

bclayman-sq opened the pull request:
Speed up the ContainExactly matcher by pre-emptively matching up corresponding elements in the expected and actual arrays. 

This addresses issues rspec/rspec#84, rspec/rspec#136.

**Before this PR, comparing two arrays of 50 random integers never finished.  Now, comparing arrays of 10,000 random integers takes 1 second.**

This PR is a collaboration between @genehsu and I based on a couple of our earlier PRs and discussion that resulted:
1) https://github.com/rspec/rspec-expectations/pull/1325
2) https://github.com/rspec/rspec-expectations/pull/1328

Co-authored-by: Gene Hsu <gene@hsufarm.com>

bclayman-sq committed:
Speed up the ContainExactly matcher

Speed up the ContainExactly matcher by pre-emptively matching up corresponding elements in the expected and actual arrays.

This addresses #1006, #1161.

This PR is a collaboration between me and @genehsu based on
a couple of our earlier PRs and discussion that resulted:
1) https://github.com/rspec/rspec-expectations/pull/1325
2) https://github.com/rspec/rspec-expectations/pull/1328

Co-authored-by: Gene Hsu (@genehsu)

genehsu commented:
@bclayman-sq 

Looks good. I'd suggest that 1000 random elements should be enough since we're hitting timing issues on one of the Jruby test suites.

> @genehsu, Could I get your email so I can update the "co-authored-by" section of the PR description?

gene `at` hsufarm `dot` com

I'm closing my previous PR now

bclayman-sq commented:
> @bclayman-sq
> 
> Looks good. I'd suggest that 1000 random elements should be enough since we're hitting timing issues on one of the Jruby test suites.
> 
> > @genehsu, Could I get your email so I can update the "co-authored-by" section of the PR description?
> 
> gene `at` hsufarm `dot` com
> 
> I'm closing my previous PR now

Ah gotcha, I ended up bumping to 2s to get around that pesky jruby one but agree in general that 1000 is probably plenty 👍 

Sounds good, will update the PR description!

genehsu commented on spec/rspec/matchers/built_in/contain_exactly_spec.rb:
```diff
@@ -184,22 +184,74 @@ def array.send; :sent; end
+      end
+    end
+
+    let(:max_runtime) { 2 }
+    let(:actual) { Array.new(10_000) { rand(10) } }
```
Suggestion: 1000 elements should be enough to see if the algorithm update works well. Then `max_runtime` can be dialed down as well, maybe back to 0.5.

bclayman-sq replied:
Ah sure thing, just updated to 1K elements and 0.5s timeout 👍

bclayman-sq committed:
Decrease number of elements in actual, expected to 1,000

JonRowe commented on spec/rspec/matchers/built_in/contain_exactly_spec.rb:
```diff
@@ -184,22 +184,74 @@ def array.send; :sent; end
-    timeout_if_not_debugging(0.1) do
-      expected = [0, 1, 1,    3, 3, 3,    4, 4,    8, 8, 9   ]
-      actual   = [   1,    2, 3, 3, 3, 3,       7, 8, 8, 9, 9]
+    shared_examples "succeeds fast" do
```
```suggestion
    shared_examples "for succeeding quickly" do
```

JonRowe commented on spec/rspec/matchers/built_in/contain_exactly_spec.rb:
```diff
@@ -184,22 +184,74 @@ def array.send; :sent; end
-      expect {
-        expect(actual).to contain_exactly(*expected)
-      }.to fail_including("the missing elements were:      [0, 1, 4, 4]")
+    shared_examples "fails fast" do |failure_msg|
```
```suggestion
    shared_examples "for failing fast with" do |failure_msg|
```

JonRowe commented on spec/rspec/matchers/built_in/contain_exactly_spec.rb:
```diff
@@ -184,22 +184,74 @@ def array.send; :sent; end
+      context "that is not valid" do
+        let(:expected) { actual.shuffle }
+
+        include_examples "fails fast", "not to contain exactly"
```
```suggestion
        include_examples "for failing fast with", "not to contain exactly"
```

JonRowe commented on spec/rspec/matchers/built_in/contain_exactly_spec.rb:
```diff
@@ -184,22 +184,74 @@ def array.send; :sent; end
+      context "that is not valid" do
+        let(:expected) { Array.new(1000) { rand(10) } }
+
+        include_examples "fails fast", "expected collection contained"
```
```suggestion
        include_examples "for failing fast with", "expected collection contained"
```

JonRowe commented on spec/rspec/matchers/built_in/contain_exactly_spec.rb:
```diff
@@ -184,22 +184,74 @@ def array.send; :sent; end
+          subject
+        end
+
+        include_examples "succeeds fast"
```
```suggestion
        include_examples "for succeeding quickly"
```

JonRowe reviewed (changes requested):
Great work, can we get a benchmark of the old implementation vs the new implementation added as a comment above the implementation to explain why its so complex, plus I have some wording tweaks for the tests as suggestion, especially the subject bit as we prefer named subjects where possible.

bclayman-sq committed:
Incorporate feedback: Add benchmark comment; use named subjects where possible

bclayman-sq commented:
> Great work, can we get a benchmark of the old implementation vs the new implementation added as a comment above the implementation to explain why its so complex, plus I have some wording tweaks for the tests as suggestion, especially the subject bit as we prefer named subjects where possible.

This is really helpful feedback; I just:
1. Added comment with benchmark data to justify the complexity
2. Incorporated the wording tweaks
3. Added named subjects where possible

One note on 3.: I didn't change the shared examples to use `expect_actual_to_contain_exactly_expected` instead of `subject`.  The specs use shared examples for both positive and negative expectations, so changing either shared example to use `expect_actual_to_contain_exactly_expected` or `expect_actual_not_to_contain_exactly_expected` wouldn't exercise the behavior we want and would cause a couple failing tests.

Let me know if I've misunderstood anything here and thanks for the review!

pirj commented on spec/rspec/matchers/built_in/contain_exactly_spec.rb:
```diff
@@ -184,22 +184,74 @@ def array.send; :sent; end
+      it do
+        timeout_if_not_debugging(max_runtime) do
+          expect {
+            subject
```
`subject` is a substitute for `expect_actual_to_contain_exactly_expected` or `expect_actual_not_to_contain_exactly_expected` depending on the including example group.

pirj reviewed (approved):
Fantastic job! Thank you

bclayman-sq commented:
@JonRowe @pirj Hi, hope you two are doing well!  I've added a comment with benchmark data per Jon's request and wanted to see if there was anything I could do to get this PR merged in.  Let me know!

(I have a couple other PRs from several months ago that I'll comment on similarly!)

bclayman-sq commented:
> 

Hi @JonRowe, @pirj has approved this and I've managed to incorporate your requested changes (benchmark data, wording tweaks).  I followed up a month or so ago and imagine you're slammed.  I do still want to get this in, however!  Would you feel OK merging this?

JonRowe commented:
Migrated to the monorepo as rspec/rspec#127
